compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -127,6 +133,36 @@ class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
           val selClassNonFinal = selClass && !(selector.typeSymbol is Final)
           val selFinalClass    = selClass && (selector.typeSymbol is Final)
 
+          /** Check if the selector's potential type parameters will be erased, and if so warn */
+          val selTypeParam = tree.args.head.tpe.widen match {
+            case tp @ AppliedType(_, arg :: _) =>
+              // If the type is `Array[X]` where `X` extends AnyVal
+              val anyValArray = tp.isRef(defn.ArrayClass) && arg.derivesFrom(defn.AnyValClass)
+              // param is: Any | AnyRef | java.lang.Object
+              val topType = defn.ObjectType <:< arg
+              // has @unchecked annotation to suppress warnings
+              val hasUncheckedAnnot = arg.hasAnnotation(defn.UncheckedAnnot)
```

Comment:
The `anyValArray` check is to make sure we allow: `xs.isInstanceOf[Array[X]]` where `X` is a primitive. However, before it would have allowed any value class - I constrained it to only allow checking on primitives.

The `topType` check is to allow explicit checking on top types, eg: 

- `xs.isInstanceOf[Array[Any]]`
- `xs.isInstanceOf[Array[AnyRef]]`
- `xs.isInstanceOf[Array[Object]]`